packages/helia/src/block-providers/trustless-gateway-block-provider.ts

```diff
@@ -0,0 +1,83 @@
+import { logger } from '@libp2p/logger'
+import type { BlockProvider } from '@helia/interface/blocks'
+import type { AbortOptions } from 'interface-store'
+import type { CID } from 'multiformats/cid'
+import type { ProgressEvent, ProgressOptions } from 'progress-events'
+
+const log = logger('helia:trustless-gateway-block-provider')
+
+export type TrustlessGatewayGetBlockProgressEvents =
+  ProgressEvent<'trustless-gateway:get-block:fetch', URL>
+
+/**
+ * A BlockProvider that accepts a list of trustless gateways that are queried
+ * for blocks. Individual gateways are randomly chosen.
+ */
+export class TrustedGatewayBlockProvider implements BlockProvider<
+ProgressOptions,
+ProgressOptions<TrustlessGatewayGetBlockProgressEvents>
+> {
+  private readonly gateways: URL[]
```

Comment:
```suggestion
class InternalGateways {
	url: URL
	attempts: number
	successes: number
	constructor(url: string|URL) {
	  this.url = url instanceOf URL ? url : url.toString()
	}
	toString() {
	  return this.url.toString()
	}
	get successRate() {
	  if (this.attempts === 0) return 0
	  return this.successes / this.attempts
	}
}

export type TrustlessGatewayGetBlockProgressEvents =
  ProgressEvent<'trustless-gateway:get-block:fetch', URL>

/**
 * A BlockProvider that accepts a list of trustless gateways that are queried
 * for blocks. Individual gateways are randomly chosen.
 */
export class TrustedGatewayBlockProvider implements BlockProvider<
ProgressOptions,
ProgressOptions<TrustlessGatewayGetBlockProgressEvents>
> {
  private readonly gateways: InternalGateways[]
```

We should have some sort of ranking/tracking of gateway retrieval attempts so we can deprioritize gateways that are consistently failing

Reply:
i'll do a PR with a slight refactor in this direction

Reply:
Sounds good.